tools/cellxgene_census_builder/src/cellxgene_census_builder/__main__.py

```diff
@@ -52,17 +52,24 @@ def do_build(args: CensusBuildArgs, skip_completed_steps: bool = False) -> int:
     logging.info(f"Census build: start [version={__version__}]")
     logging.info(args)
 
-    build_steps: List[Callable[[CensusBuildArgs], bool]] = [
-        do_prebuild_set_defaults,
-        do_prebuild_checks,
-        do_build_soma,
-        do_validate_soma,
-        do_create_reports,
-        do_data_copy,
-        do_the_release,
-        do_report_copy,
-        do_old_release_cleanup,
-    ]
+    # Infer steps from environment variables, if provided
+    if os.getenv("CENSUS_BUILD_STEPS"):
```

Comment:
this model makes me concerned that we could easily get steps out of order (i.e., the external code could get out of sync with the _required_ execution order).

I would be more comfortable if each step could explicitly check that its dependencies were fulfilled.  The old code did this implicitly with the ordered `build_steps` array, which could not be executed out of order.

The new code is (inferring steps from environment) is fragile, as it allows execution in any order, skipping steps, etc.  The entire system assumes this can't happen, so we need to protect against that. In other words, it is a state machine (simple sequence that can pause/resume), and the new code doesn't enforce that.

Concrete proposal:
1. the env variable steps should be sanity checked for order and that they don't skip anything (e.g.,  step A, B, D when it should be A, B, C, D).
2. each individual step does at least a reasonable check that its upstream dependent steps were executed (via the state and possibly more)

You could skip the first item if you simply let the environment say "start at step X", and didn't allow multiple (possibly out of order) execution.

TL;DR - I'd like to see this get substantially hardened before we deploy it.

Reply:
I think the best solution here is to use the `build-state.yaml` file to ensure that each step can start only if all the pre-requisite steps have successfully been completed. Due to the nature of the AWS SFN (and potentially other runners like Airflow), the state needs to be saved in a shared location, e.g. S3. This is a medium effort change, so before I implement it I would make sure we all agree on it. Thoughts @bkmartinjr @atolopko-czi?

Reply:
I agree that it should use build-state (the code already records each step as having been completed).  It already records the succexs of each step with `args.state[build_step.__name__] = True`

Assuming all steps are idempotent (which I _think_ is true), you could simply add a check that the immediately preceding step was completed successfully.

I.e., I'm thinking something like:
1. in `do_build`, make sure you set the step completion state to False before each step (it currently assumes steps will never re-run)
2. In each `do_step_B`, which depends on on `do_step_A`, add something like `assert args.state['do_step_A']`

Reply:
The issue is that the execution will now be split into multiple containers so we can't do this in memory. In particular, the `sync` step (the new one) and the `release` will run in separate batch jobs.

Reply:
It already commits it to the state file (see line 79 in __main__.py), so the "in file" solution is already in place.  I.e., it was built as a "not" memory only solutoin.

We should talk in person if lines 78, 79 are not clear.

Reply:
I see the commit and I think it's pretty much what we need, aside from:
1. Adding the state file to S3 so it can be used by separate containers
2. Adding explicit checks to make sure that steps can only execute in the right state.
Happy to chat in person if you want to speed up the decision.